File: tests/test_vsa.py
"""PHYSMOL VSA Python Integration Tests."""

import numpy as np
import pytest
import sys
import os

# Add src to path for development
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src', 'python'))

from physmol.vsa import VectorSymbolicArchitecture, VSAVector, Codebook, FHRRSpace


class TestVSAVector:
    def test_create(self):
        vsa = VectorSymbolicArchitecture(4096)
        v = vsa.random_bipolar()
        assert v.dim == 4096
        assert len(v.data) == 4096
        assert v.data.dtype == np.float32

    def test_bind(self):
        vsa = VectorSymbolicArchitecture(4096)
        a = vsa.random_bipolar(1)
        b = vsa.random_bipolar(2)
        c = vsa.bind(a, b)
        # Verify Hadamard product
        expected = a.data * b.data
        np.testing.assert_allclose(c.data, expected, atol=1e-6)

    def test_bind_commutativity(self):
        vsa = VectorSymbolicArchitecture(4096)
        a = vsa.random_bipolar(10)
        b = vsa.random_bipolar(20)
        ab = vsa.bind(a, b)
        ba = vsa.bind(b, a)
        np.testing.assert_allclose(ab.data, ba.data, atol=1e-6)

    def test_bundle(self):
        vsa = VectorSymbolicArchitecture(4096)
        a = vsa.random_bipolar(3)
        b = vsa.random_bipolar(4)
        c = vsa.bundle(a, b)
        expected = a.data + b.data
        np.testing.assert_allclose(c.data, expected, atol=1e-6)

    def test_similarity_self(self):
        vsa = VectorSymbolicArchitecture(4096)
        a = vsa.random_bipolar(5)
        sim = vsa.similarity(a, a)
        assert abs(sim - 1.0) < 1e-4

    def test_similarity_random(self):
        vsa = VectorSymbolicArchitecture(4096)
        a = vsa.random_bipolar(100)
        b = vsa.random_bipolar(200)
        sim = vsa.similarity(a, b)
        assert abs(sim) < 0.1  # nearly orthogonal

    def test_permute(self):
        vsa = VectorSymbolicArchitecture(4096)
        v = vsa.random_bipolar(7)
        shifted = vsa.permute(v, 1)
        assert shifted.data[0] == v.data[1]
        assert shifted.data[-1] == v.data[0]

    def test_normalize(self):
        vsa = VectorSymbolicArchitecture(4096)
        v = vsa.random_bipolar(8)
        normed = vsa.normalize(v)
        norm = np.linalg.norm(normed.data)
        assert abs(norm - 1.0) < 1e-4

    def test_quantize(self):
        vsa = VectorSymbolicArchitecture(4096)
        v = vsa.random_bipolar(9)
        q = vsa.quantize(v, bits=8)
        # Should be close to original
        diff = np.max(np.abs(v.data - q.data))
        assert diff < 0.1


class TestCodebook:
    def test_add_lookup(self):
        cb = Codebook(4096)
        vsa = VectorSymbolicArchitecture(4096)
        red = vsa.random_bipolar(100)
        blue = vsa.random_bipolar(200)

        cb.add("red", red)
        cb.add("blue", blue)
        assert len(cb) == 2

        found = cb.lookup("red")
        assert found is not None
        np.testing.assert_allclose(found.data, red.data, atol=1e-6)

    def test_nearest(self):
        cb = Codebook(4096)
        vsa = VectorSymbolicArchitecture(4096)
        red = vsa.random_bipolar(100)
        blue = vsa.random_bipolar(200)
        cb.add("red", red)
        cb.add("blue", blue)

        idx, sim, name = cb.nearest(red)
        assert name == "red"
        assert sim > 0.99


class TestFHRRSpace:
    def test_encode_position(self):
        fhrr = FHRRSpace(4096)
        v1 = fhrr.encode_position(1.0, 2.0, 3.0)
        v2 = fhrr.encode_position(1.0, 2.0, 3.0)
        # Same position should give same vector
        np.testing.assert_allclose(v1.data, v2.data, atol=1e-6)

    def test_different_positions(self):
        fhrr = FHRRSpace(4096)
        v1 = fhrr.encode_position(0, 0, 0)
        v2 = fhrr.encode_position(1, 0, 0)
        vsa = VectorSymbolicArchitecture(4096)
        sim = vsa.similarity(v1, v2)
        # Different positions should be different
        assert sim < 0.99

    def test_translate(self):
        fhrr = FHRRSpace(4096)
        v1 = fhrr.encode_position(0, 0, 0)
        v2 = fhrr.translate(v1, 1, 0, 0)
        v3 = fhrr.encode_position(1, 0, 0)
        vsa = VectorSymbolicArchitecture(4096)
        # Translation should produce similar result to direct encoding
        sim = vsa.similarity(v2, v3)
        # They won't be identical due to binding noise, but should be correlated
        assert sim > 0.0


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/test_vsa.c
/* PHYSMOL VSA C Unit Tests */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include "../src/core/vsa.h"

#define DIM 4096
#define TOLERANCE 1e-4f

static int test_count = 0;
static int pass_count = 0;

#define TEST(name) do { \
    printf("  TEST: %-40s ", name); \
    test_count++; \
} while(0)

#define PASS() do { printf("[PASS]\n"); pass_count++; } while(0)
#define FAIL(msg) do { printf("[FAIL] %s\n", msg); } while(0)

void test_create_free() {
    TEST("create and free vector");
    VSAVector *v = vsa_create(DIM);
    assert(v != NULL);
    assert(v->dim == DIM);
    for (size_t i = 0; i < DIM; i++) assert(v->data[i] == 0.0f);
    vsa_free(v);
    PASS();
}

void test_random_bipolar() {
    TEST("random bipolar generation");
    VSAVector *v = vsa_create(DIM);
    vsa_random_bipolar(v, 42);
    int has_pos = 0, has_neg = 0;
    for (size_t i = 0; i < DIM; i++) {
        if (v->data[i] == 1.0f) has_pos = 1;
        else if (v->data[i] == -1.0f) has_neg = 1;
        else { FAIL("non-bipolar value found"); return; }
    }
    assert(has_pos && has_neg);
    vsa_free(v);
    PASS();
}

void test_bind() {
    TEST("binding (Hadamard product)");
    VSAVector *a = vsa_create(DIM);
    VSAVector *b = vsa_create(DIM);
    VSAVector *out = vsa_create(DIM);
    vsa_random_bipolar(a, 1);
    vsa_random_bipolar(b, 2);
    vsa_bind(out, a, b);
    for (size_t i = 0; i < DIM; i++) {
        float expected = a->data[i] * b->data[i];
        assert(fabsf(out->data[i] - expected) < TOLERANCE);
    }
    vsa_free(a); vsa_free(b); vsa_free(out);
    PASS();
}

void test_bind_commutativity() {
    TEST("binding commutativity: a*b == b*a");
    VSAVector *a = vsa_create(DIM);
    VSAVector *b = vsa_create(DIM);
    VSAVector *ab = vsa_create(DIM);
    VSAVector *ba = vsa_create(DIM);
    vsa_random_bipolar(a, 10);
    vsa_random_bipolar(b, 20);
    vsa_bind(ab, a, b);
    vsa_bind(ba, b, a);
    for (size_t i = 0; i < DIM; i++) {
        assert(fabsf(ab->data[i] - ba->data[i]) < TOLERANCE);
    }
    vsa_free(a); vsa_free(b); vsa_free(ab); vsa_free(ba);
    PASS();
}

void test_bundle() {
    TEST("bundling (addition)");
    VSAVector *a = vsa_create(DIM);
    VSAVector *b = vsa_create(DIM);
    VSAVector *out = vsa_create(DIM);
    vsa_random_bipolar(a, 3);
    vsa_random_bipolar(b, 4);
    vsa_bundle(out, a, b);
    for (size_t i = 0; i < DIM; i++) {
        float expected = a->data[i] + b->data[i];
        assert(fabsf(out->data[i] - expected) < TOLERANCE);
    }
    vsa_free(a); vsa_free(b); vsa_free(out);
    PASS();
}

void test_cosine_similarity() {
    TEST("cosine similarity");
    VSAVector *a = vsa_create(DIM);
    VSAVector *b = vsa_create(DIM);
    vsa_random_bipolar(a, 5);
    /* Self-similarity should be 1.0 */
    float sim_aa = vsa_cosine_similarity(a, a);
    assert(fabsf(sim_aa - 1.0f) < TOLERANCE);
    /* Random vectors should be nearly orthogonal */
    vsa_random_bipolar(b, 6);
    float sim_ab = vsa_cosine_similarity(a, b);
    assert(fabsf(sim_ab) < 0.1f); /* within 0.1 with high probability */
    vsa_free(a); vsa_free(b);
    PASS();
}

void test_quasi_orthogonality() {
    TEST("quasi-orthogonality of random vectors");
    int num_vecs = 100;
    VSAVector **vecs = (VSAVector **)malloc(num_vecs * sizeof(VSAVector *));
    for (int i = 0; i < num_vecs; i++) {
        vecs[i] = vsa_create(DIM);
        vsa_random_bipolar(vecs[i], i + 100);
    }
    float max_sim = 0.0f;
    int count_above_threshold = 0;
    for (int i = 0; i < num_vecs; i++) {
        for (int j = i + 1; j < num_vecs; j++) {
            float sim = fabsf(vsa_cosine_similarity(vecs[i], vecs[j]));
            if (sim > max_sim) max_sim = sim;
            if (sim > 0.1f) count_above_threshold++;
        }
    }
    /* With D=4096, σ=1/√4096≈0.016, so |S|>0.1 should be rare */
    printf("(max_sim=%.4f, pairs_above_0.1=%d) ", max_sim, count_above_threshold);
    assert(max_sim < 0.15f);
    for (int i = 0; i < num_vecs; i++) vsa_free(vecs[i]);
    free(vecs);
    PASS();
}

void test_permute() {
    TEST("circular permutation");
    VSAVector *v = vsa_create(DIM);
    VSAVector *out = vsa_create(DIM);
    vsa_random_bipolar(v, 7);
    vsa_permute(out, v, 1);
    /* Shifted by 1: out[0] == v[1], out[DIM-1] == v[0] */
    assert(fabsf(out->data[0] - v->data[1]) < TOLERANCE);
    assert(fabsf(out->data[DIM - 1] - v->data[0]) < TOLERANCE);
    vsa_free(v); vsa_free(out);
    PASS();
}

void test_normalize() {
    TEST("normalize to unit length");
    VSAVector *v = vsa_create(DIM);
    vsa_random_bipolar(v, 8);
    vsa_normalize(v);
    float norm_sq = 0.0f;
    for (size_t i = 0; i < DIM; i++) norm_sq += v->data[i] * v->data[i];
    assert(fabsf(norm_sq - 1.0f) < TOLERANCE);
    vsa_free(v);
    PASS();
}

void test_codebook() {
    TEST("codebook add/lookup/nearest");
    VSACodebook *cb = vsa_codebook_create(DIM, 16);
    VSAVector *red = vsa_create(DIM);
    VSAVector *blue = vsa_create(DIM);
    VSAVector *green = vsa_create(DIM);
    vsa_random_bipolar(red, 100);
    vsa_random_bipolar(blue, 200);
    vsa_random_bipolar(green, 300);

    vsa_codebook_add(cb, "red", red);
    vsa_codebook_add(cb, "blue", blue);
    vsa_codebook_add(cb, "green", green);
    assert(cb->num_vectors == 3);

    /* Lookup */
    const VSAVector *found = vsa_codebook_lookup(cb, "red");
    assert(found != NULL);

    /* Nearest: query with red, should find red */
    float sim;
    size_t idx = vsa_codebook_nearest(cb, red, &sim);
    assert(idx == 0); /* red was added first */
    assert(sim > 0.99f);

    vsa_free(red); vsa_free(blue); vsa_free(green);
    vsa_codebook_free(cb);
    PASS();
}

void test_fpe_encode() {
    TEST("FPE spatial encoding");
    VSAVector *out = vsa_create(DIM);
    VSAVector *bx = vsa_create(DIM);
    VSAVector *by = vsa_create(DIM);
    VSAVector *bz = vsa_create(DIM);
    vsa_random_phase(bx, 500);
    vsa_random_phase(by, 600);
    vsa_random_phase(bz, 700);

    vsa_fpe_encode(out, 1.0f, 2.0f, 3.0f, bx, by, bz);

    /* Verify: phase should be x*bx + y*by + z*bz */
    for (size_t i = 0; i < DIM; i++) {
        float expected = 1.0f * bx->data[i] + 2.0f * by->data[i] + 3.0f * bz->data[i];
        assert(fabsf(out->data[i] - expected) < TOLERANCE);
    }

    vsa_free(out); vsa_free(bx); vsa_free(by); vsa_free(bz);
    PASS();
}

int main() {
    printf("=== PHYSMOL VSA Unit Tests (DIM=%d) ===\n", DIM);

    test_create_free();
    test_random_bipolar();
    test_bind();
    test_bind_commutativity();
    test_bundle();
    test_cosine_similarity();
    test_quasi_orthogonality();
    test_permute();
    test_normalize();
    test_codebook();
    test_fpe_encode();

    printf("\n%d/%d tests passed\n", pass_count, test_count);
    return (pass_count == test_count) ? 0 : 1;
}
